### ISSUE TITLE
Lint Suboptimal Expression Order

### ISSUE THREAD

**Issue by llogiq** [Manishearth/rust-clippy#1193]
(This one is nicked from fb-contrib). Lint expressions of the kind

(method)call &&/|| (no method call) {==, !=} (no method call)

e.g. `potentially_expensive_check() && a == b`

And suggest to reorder the subexpressions to possibly improve performance. Note that this may change behavior if `potentially_expensive_check()` has side effects.

**Comment by nk9** [comment]
> Note that this may change behavior if `potentially_expensive_check()` has side effects.

Indeed, if the method call sets some state, then the subsequent comparison could be impacted. I'm not sure there's a way to figure out when to invoke this check purely through static analysis.

**Comment by mi2ebi** [comment]
This might also break code that relies on short circuiting
